src/backend.c

```diff
@@ -305,34 +318,143 @@ do_state_check (int minlvl, const char *where)
+            /* Do nothing. */
+            break;
+
+        case DCS_TERMINATE:
+        {
+            /* Call the default handler, for all four signals
+             * this is termination.
+             */
+            struct sigaction sa;
+            sigemptyset(&sa.sa_mask);
+            sa.sa_flags = 0;
+            sa.sa_handler = SIG_DFL;
+            if (sigaction(sig, &sa, NULL) == 0)
+                raise(sig);
+            /* If we are still here, exit... */
+            exit(sig);
+            return;
+        }
+
+        case DCS_THROW_EXCEPTION:
+            /* Set a flag to be evaluated by the interpreter. */
+            interrupt_execution = true;
+#ifdef USE_PYTHON
+            /* Notify the python package. */
+            python_interrupt();
+#endif
+            break;
+
+        default:
+            /* Process it in the backend loop. */
+            sigaddset(&pending_signals, sig);
+            extra_jobs_to_do = MY_TRUE;
+            return;
+    }
+
+    if (sig == SIGINT)
+        install_sigint_handler();
+
 } /* handle_signal() */
 
 /*-------------------------------------------------------------------------*/
-static INLINE Bool
-defer_signal_to_master(int sig)
-// Notifies master about the signal <sig>.
-// Returns 1 if the master returns anything != 0. In this case it is assumed,
-// the master handled the signal.
+static bool
+install_sigint_handler ()
+
+/* Installs the SIGINT handler again.
+ * The handler is reset upon each signal, so we need to install it again.
+ * Returns true on success.
+ */
+
+{
+    struct sigaction sa;
+    sigemptyset(&sa.sa_mask);
+    sa.sa_flags = SA_RESTART|SA_RESETHAND;
+    sa.sa_handler = handle_signal;
+    return (sigaction(SIGINT, &sa, NULL) == 0);
+}
+
+/*-------------------------------------------------------------------------*/
+static INLINE bool
+execute_signal_action (int sig, char configured)
+
+/* Executes the configured signal action, that were not already handled in the
+ * signal handler itself: DEFAULT, SHUTDOWN, CALL_MASTER and RELOAD_MASTER.
+ *
+ * Returns true, if the action was executed.
+ * Returns false, if the default action (besides calling the master)
+ * shall be conducted.
+ */
+
 {
-    svalue_t *res;
-    push_number(inter_sp, sig);
-    res = callback_master(STR_HANDLE_EXTERNAL_SIGNAL, 1);
-    if (res && res->type == T_NUMBER && res->u.number != 0)
+    switch (configured)
     {
-        return MY_TRUE;
+        case DCS_SHUTDOWN:
+            extra_jobs_to_do = MY_TRUE;
+            game_is_being_shut_down = MY_TRUE;
+            break;
+
+        case DCS_RELOAD_MASTER:
+            extra_jobs_to_do = MY_TRUE;
+            master_will_be_updated = MY_TRUE;
+            break;
+
+        case DCS_CALL_MASTER:
+        default:
+        {
+            svalue_t *res;
+
+            push_number(inter_sp, sig);
+            res = callback_master(STR_HANDLE_EXTERNAL_SIGNAL, 1);
+
+            if (configured != DCS_CALL_MASTER &&
```

Comment:
Das Konstrukt finde ich echt schwer lesbar. 
Vielleicht liegts an dem Knoten im Hirn, aber ich erwarte, dass bei configured == DCS_CALL_MASTER immer ein false zurueckgeben wird, aber in dem Fall kommen wir weiter unten ins return true...?

Reply:
War tatsächlich fehlerhaft. Hab es jetzt kommentiert.